doc/fast-reboot/Fast-reboot_Flow_Improvements_HLD.md

```diff
@@ -0,0 +1,158 @@
+* [Neighbors](#43-neighsyncd-point-of-view)
+* [BGP](#44-fpmsyncd-point-of-view)
+* [Finalizer](#45-reboot-finalizer)
+
+# 1 Overview
+
+The goal of SONiC fast-reboot is to be able to restart and upgrade SONiC software with a data plane disruption less than 30 seconds and control plane less than 90 seconds.
+With current implementation there is no indication of the fast-reboot status, meaning we don't have a way to determine if the flow has finished or not.
+Some feature flows in SONiC are delayed with a timer to keep the CPU dedicated to the fast-reboot init flow for best perforamnce, like enablement of flex counters.
+In order to have such indicator, re-use of the fastfast-reboot infrastructure can be used.
+
+Each network application will experience similar processing flow.
+Application and corresponding orchagent sub modules need to work together to restore the original data and push it to the ASIC.
+Take neighbor as an example, upon restart operation every neighbor we had prior to the reboot should be created again after resetting the ASIC.
+We should also synchronize the actual neighbor state after recovering it, the MAC of the neighbor could have changed, went down for some reason etc.
+In this case, restore_neighbors.py script will align the network state with the switch state by sending ARP/NDP to all known neighbors prior the reboot.
+neighsyncd will update the internal cache with all neighbors and push all to APP DB, orchagent will then add/remove/update any neighbor and get syncd to program the HW with the new data.
+
+In addition to the recover mechanism, the warmboot-finalizer can be enhanced to finalize fast-reboot as well and introduce a new flag indicating the process is done.
+This new flag can be used later on for any functionality which we want to start only after init flow finished in case of fast-reboot.
+This is to prevent interference in the fast-reboot reconciliation process and impair the performance, for example enablement of flex counters.
+
+References:
+https://github.com/Azure/SONiC/blob/master/doc/fast-reboot/fastreboot.pdf
+https://github.com/Azure/sonic-swss/blob/master/neighsyncd/restore_neighbors.py
+https://github.com/Azure/sonic-buildimage/blob/master/dockers/docker-orchagent/enable_counters.py
+
+# 2 Functional Requirements
+
+The new Fast-reboot design should meet the following requirments:
+
+- Reboot the switch into a new SONiC software version using kexec.
+- Upgrade the FW by the new SONiC image if needed.
+- Recover all applications state with the new image to the previous state prior the reboot.
+- Recover ASIC state after reset to the previous state prior the reboot.
+- Recover the Kernel state after reset to the previous state prior the reboot.
+- Sync the Kernel and ASIC with changes on the network which happen during fast-reboot.
+- Control plane downtime will not exceed 90 seconds.
+- Data plane downtime will not exceed 30 seconds.
```

Comment:
I thought it will be better to gather all requirements, even the ones which are already satisfied with previous implementation since we are changing it to a new one.
I think it is important to mention all when the implementation is different.
What do you think?